Supporting API levels < 18  by using an alternative to UsbDeviceConnection.bulkTransfer

bulkTransfer method used for device communications currently requires a minimum API level of 18.
For API levels <18, another method that uses write function on the device after checking the API version may be added.
The present code crashes on an API 15 device : Simmtronics XPAD 1010 10" tablet with the following runtime error:
```
06-11 19:16:43.584 2168-2168/org.fossasia.pslab W/dalvikvm: threadid=1: thread exiting with uncaught exception (group=0x40a121f8)
06-11 19:16:43.594 2168-2168/org.fossasia.pslab E/AndroidRuntime: FATAL EXCEPTION: main
                                                                  java.lang.NoSuchMethodError: android.hardware.usb.UsbDeviceConnection.bulkTransfer
                                                                      at org.fossasia.pslab.communication.CommunicationHandler.write(CommunicationHandler.java:149)
                                                                      at org.fossasia.pslab.communication.PacketHandler.sendByte(PacketHandler.java:60)
                                                                      at org.fossasia.pslab.communication.PacketHandler.getVersion(PacketHandler.java:43)
                                                                      at org.fossasia.pslab.communication.ScienceLab.getVersion(ScienceLab.java:431)
                                                                      at org.fossasia.pslab.activity.MainActivity.getHomeFragment(MainActivity.java:147)
                                                                      at org.fossasia.pslab.activity.MainActivity.access$000(MainActivity.java:46)
                                                                      at org.fossasia.pslab.activity.MainActivity$1.run(MainActivity.java:118)
                                                                      at android.os.Handler.handleCallback(Handler.java:605)
```

@viveksb007 
I believe there exists a permissions issue because the app successfully reads the version string sometimes.
How to reproduce this : 
 - Connect the PSLab, and the OS (Kitkat 4.0.4 APi 15) automatically opens a dialog asking if the PSLab app should be opened. In this case, the version is detected correctly . no problem.  Error log 1
- Launch the app manually, and it promptly crashes. Error log 2
- I could not reproduce the behaviour where the app launches successfully but the version string is not displayed.

<details><summary>Error log 1 : app auto-launched. works fine</summary><p>

```
06-15 12:29:54.110 1187-1191/org.fossasia.pslab I/dalvikvm: threadid=3: reacting to signal 3
06-15 12:29:54.130 1187-1191/org.fossasia.pslab I/dalvikvm: Wrote stack traces to '/data/anr/traces.txt'
06-15 12:29:54.130 1187-1187/org.fossasia.pslab D/CommunicationHandler: VID : 3034PID : 33142
06-15 12:29:54.140 1187-1187/org.fossasia.pslab D/CommunicationHandler: VID : 1240PID : 223
06-15 12:29:54.140 1187-1187/org.fossasia.pslab D/CommunicationHandler: Found PSLAB Device
06-15 12:29:54.140 1187-1187/org.fossasia.pslab D/CommunicationHandler: Claiming interfaces, count=3
06-15 12:29:54.140 1187-1187/org.fossasia.pslab D/CommunicationHandler: Control interface=UsbInterface[mId=0,mClass=2,mSubclass=2,mProtocol=1,mEndpoints=[Landroid.os.Parcelable;@4105dc80]
06-15 12:29:54.140 1187-1187/org.fossasia.pslab D/CommunicationHandler: Control endpoint direction: 128
06-15 12:29:54.140 1187-1187/org.fossasia.pslab D/CommunicationHandler: Claiming data interface.
06-15 12:29:54.150 1187-1187/org.fossasia.pslab D/CommunicationHandler: data interface=UsbInterface[mId=1,mClass=10,mSubclass=0,mProtocol=0,mEndpoints=[Landroid.os.Parcelable;@4108e120]
06-15 12:29:54.150 1187-1187/org.fossasia.pslab D/CommunicationHandler: Read endpoint direction: 128
06-15 12:29:54.150 1187-1187/org.fossasia.pslab D/CommunicationHandler: Write endpoint direction: 0
06-15 12:29:54.350 1187-1187/org.fossasia.pslab D/MainActivity: PSLab device found
06-15 12:29:54.350 1187-1187/org.fossasia.pslab I/dalvikvm: Could not find method android.widget.ImageButton.hasOverlappingRendering, referenced from method android.support.v7.widget.AppCompatImageButton.hasOverlappingRendering
06-15 12:29:54.350 1187-1187/org.fossasia.pslab W/dalvikvm: VFY: unable to resolve virtual method 23317: Landroid/widget/ImageButton;.hasOverlappingRendering ()Z
06-15 12:29:54.350 1187-1187/org.fossasia.pslab D/dalvikvm: VFY: replacing opcode 0x6f at 0x0008
06-15 12:29:54.380 1187-1187/org.fossasia.pslab E/dalvikvm: Could not find class 'android.graphics.drawable.RippleDrawable', referenced from method android.support.v7.widget.AppCompatImageHelper.hasOverlappingRendering
06-15 12:29:54.380 1187-1187/org.fossasia.pslab W/dalvikvm: VFY: unable to resolve instanceof 160 (Landroid/graphics/drawable/RippleDrawable;) in Landroid/support/v7/widget/AppCompatImageHelper;
06-15 12:29:54.380 1187-1187/org.fossasia.pslab D/dalvikvm: VFY: replacing opcode 0x20 at 0x000c
06-15 12:29:54.410 1187-1187/org.fossasia.pslab D/ViewRootImpl: pckname = org.fossasia.pslab
06-15 12:29:54.420 1187-1187/org.fossasia.pslab V/CommunicationHandler: Read something[B@41164cf8
06-15 12:29:54.420 1187-1187/org.fossasia.pslab V/CommunicationHandler: Read something[B@41164cf8
06-15 12:29:54.610 1187-1191/org.fossasia.pslab I/dalvikvm: threadid=3: reacting to signal 3
06-15 12:29:54.620 1187-1191/org.fossasia.pslab I/dalvikvm: Wrote stack traces to '/data/anr/traces.txt'
06-15 12:29:54.920 1187-1187/org.fossasia.pslab E/CommunicationHandler: Read Error: 14
06-15 12:29:54.940 1187-1187/org.fossasia.pslab D/libEGL: loaded /system/lib/egl/libEGL_mali.so
06-15 12:29:54.950 1187-1187/org.fossasia.pslab D/libEGL: loaded /system/lib/egl/libGLESv1_CM_mali.so
06-15 12:29:54.960 1187-1187/org.fossasia.pslab D/libEGL: loaded /system/lib/egl/libGLESv2_mali.so
06-15 12:29:55.000 1187-1187/org.fossasia.pslab D/OpenGLRenderer: Enabling debug mode 0
06-15 12:29:55.010 1187-1187/org.fossasia.pslab I/dalvikvm: Could not find method android.support.v7.widget.LinearLayoutCompat.drawableHotspotChanged, referenced from method android.support.design.internal.ForegroundLinearLayout.drawableHotspotChanged
06-15 12:29:55.020 1187-1187/org.fossasia.pslab W/dalvikvm: VFY: unable to resolve virtual method 19281: Landroid/support/v7/widget/LinearLayoutCompat;.drawableHotspotChanged (FF)V
06-15 12:29:55.020 1187-1187/org.fossasia.pslab D/dalvikvm: VFY: replacing opcode 0x6f at 0x0000
06-15 12:29:55.040 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.040 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.050 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.050 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.050 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.050 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.050 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.050 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.060 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.070 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.070 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.070 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.070 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.070 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.070 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.070 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.080 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.080 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.080 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.080 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.090 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.090 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.090 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.090 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.100 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.100 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.100 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.100 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.100 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.100 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.100 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.100 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.110 1187-1191/org.fossasia.pslab I/dalvikvm: threadid=3: reacting to signal 3
06-15 12:29:55.120 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.120 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.120 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.130 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.130 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.130 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.130 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.130 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.140 1187-1191/org.fossasia.pslab I/dalvikvm: Wrote stack traces to '/data/anr/traces.txt'
06-15 12:29:55.150 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.150 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.150 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.150 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.150 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.150 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.150 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.150 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.160 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.160 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.160 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.160 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.160 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.170 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.170 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.170 1187-1187/org.fossasia.pslab W/ResourceType: Skipping entry 0x109003a in package table 0 because it is not complex!
06-15 12:29:55.180 1187-1187/org.fossasia.pslab I/dalvikvm: Could not find method android.widget.ImageView.hasOverlappingRendering, referenced from method android.support.v7.widget.AppCompatImageView.hasOverlappingRendering
06-15 12:29:55.180 1187-1187/org.fossasia.pslab W/dalvikvm: VFY: unable to resolve virtual method 23348: Landroid/widget/ImageView;.hasOverlappingRendering ()Z
06-15 12:29:55.180 1187-1187/org.fossasia.pslab D/dalvikvm: VFY: replacing opcode 0x6f at 0x0008
06-15 12:29:55.210 1187-1187/org.fossasia.pslab D/dalvikvm: GC_FOR_ALLOC freed 104K, 3% free 7684K/7879K, paused 26ms
06-15 12:29:55.210 1187-1187/org.fossasia.pslab I/dalvikvm-heap: Grow heap (frag case) to 8.534MB for 1048592-byte allocation
06-15 12:29:55.250 1187-1189/org.fossasia.pslab D/dalvikvm: GC_CONCURRENT freed 10K, 4% free 8697K/8967K, paused 1ms+2ms
06-15 12:29:55.280 1187-1187/org.fossasia.pslab D/dalvikvm: GC_FOR_ALLOC freed 2K, 3% free 8702K/8967K, paused 21ms
06-15 12:29:55.290 1187-1187/org.fossasia.pslab I/dalvikvm-heap: Grow heap (frag case) to 9.528MB for 1048592-byte allocation
06-15 12:29:55.320 1187-1187/org.fossasia.pslab D/dalvikvm: GC_FOR_ALLOC freed <1K, 4% free 9725K/10055K, paused 21ms
06-15 12:31:11.970 1187-1187/org.fossasia.pslab D/OpenGLRenderer: Flushing caches (mode 0)
06-15 12:31:12.050 1187-1187/org.fossasia.pslab D/OpenGLRenderer: Flushing caches (mode 1)
06-15 12:31:42.130 1187-1187/org.fossasia.pslab D/UsbDeviceConnectionJNI: close
```
</p></details>




<details><summary>Error log 2 : app crashes when manually launched</summary><p>

```
06-15 12:33:59.440 3203-3203/org.fossasia.pslab I/dalvikvm: Could not find method android.widget.LinearLayout$LayoutParams.<init>, referenced from method android.support.design.widget.AppBarLayout$LayoutParams.<init>
06-15 12:33:59.440 3203-3203/org.fossasia.pslab W/dalvikvm: VFY: unable to resolve direct method 23379: Landroid/widget/LinearLayout$LayoutParams;.<init> (Landroid/widget/LinearLayout$LayoutParams;)V
06-15 12:33:59.440 3203-3203/org.fossasia.pslab D/dalvikvm: VFY: replacing opcode 0x70 at 0x0000
06-15 12:33:59.500 3203-3203/org.fossasia.pslab I/dalvikvm: Could not find method android.hardware.usb.UsbDeviceConnection.bulkTransfer, referenced from method org.fossasia.pslab.communication.CommunicationHandler.write
06-15 12:33:59.500 3203-3203/org.fossasia.pslab W/dalvikvm: VFY: unable to resolve virtual method 977: Landroid/hardware/usb/UsbDeviceConnection;.bulkTransfer (Landroid/hardware/usb/UsbEndpoint;[BIII)I
06-15 12:33:59.510 3203-3203/org.fossasia.pslab D/dalvikvm: VFY: replacing opcode 0x74 at 0x0021
06-15 12:33:59.510 3203-3203/org.fossasia.pslab D/CommunicationHandler: VID : 1240PID : 223
06-15 12:33:59.510 3203-3203/org.fossasia.pslab D/CommunicationHandler: Found PSLAB Device
06-15 12:33:59.520 3203-3203/org.fossasia.pslab E/UsbManager: exception in UsbManager.openDevice
                                                              java.lang.SecurityException: User has not given permission to device UsbDevice[mName=/dev/bus/usb/003/003,mVendorId=1240,mProductId=223,mClass=239,mSubclass=2,mProtocol=1,mInterfaces=[Landroid.hardware.usb.UsbInterface;@4159b938]
                                                                  at android.os.Parcel.readException(Parcel.java:1327)
                                                                  at android.os.Parcel.readException(Parcel.java:1281)
                                                                  at android.hardware.usb.IUsbManager$Stub$Proxy.openDevice(IUsbManager.java:340)
                                                                  at android.hardware.usb.UsbManager.openDevice(UsbManager.java:255)
                                                                  at org.fossasia.pslab.communication.CommunicationHandler.open(CommunicationHandler.java:75)
                                                                  at org.fossasia.pslab.communication.ScienceLab.<init>(ScienceLab.java:81)
                                                                  at org.fossasia.pslab.others.ScienceLabCommon.openDevice(ScienceLabCommon.java:22)
                                                                  at org.fossasia.pslab.activity.MainActivity.onCreate(MainActivity.java:80)
                                                                  at android.app.Activity.performCreate(Activity.java:4492)
                                                                  at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1049)
                                                                  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:1920)
                                                                  at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:1981)
                                                                  at android.app.ActivityThread.access$600(ActivityThread.java:123)
                                                                  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1147)
                                                                  at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                  at android.os.Looper.loop(Looper.java:137)
                                                                  at android.app.ActivityThread.main(ActivityThread.java:4424)
                                                                  at java.lang.reflect.Method.invokeNative(Native Method)
                                                                  at java.lang.reflect.Method.invoke(Method.java:511)
                                                                  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:784)
                                                                  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:551)
                                                                  at dalvik.system.NativeStart.main(Native Method)
06-15 12:33:59.520 3203-3203/org.fossasia.pslab D/CommunicationHandler: Claiming interfaces, count=3
06-15 12:33:59.530 3203-3203/org.fossasia.pslab D/CommunicationHandler: Control interface=UsbInterface[mId=0,mClass=2,mSubclass=2,mProtocol=1,mEndpoints=[Landroid.os.Parcelable;@414547c8]
06-15 12:33:59.530 3203-3203/org.fossasia.pslab D/AndroidRuntime: Shutting down VM
06-15 12:33:59.530 3203-3203/org.fossasia.pslab W/dalvikvm: threadid=1: thread exiting with uncaught exception (group=0x409f91f8)
06-15 12:33:59.540 3203-3203/org.fossasia.pslab E/AndroidRuntime: FATAL EXCEPTION: main
                                                                  java.lang.RuntimeException: Unable to start activity ComponentInfo{org.fossasia.pslab/org.fossasia.pslab.activity.MainActivity}: java.lang.NullPointerException
                                                                      at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:1956)
                                                                      at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:1981)
                                                                      at android.app.ActivityThread.access$600(ActivityThread.java:123)
                                                                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1147)
                                                                      at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                      at android.os.Looper.loop(Looper.java:137)
                                                                      at android.app.ActivityThread.main(ActivityThread.java:4424)
                                                                      at java.lang.reflect.Method.invokeNative(Native Method)
                                                                      at java.lang.reflect.Method.invoke(Method.java:511)
                                                                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:784)
                                                                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:551)
                                                                      at dalvik.system.NativeStart.main(Native Method)
                                                                   Caused by: java.lang.NullPointerException
                                                                      at org.fossasia.pslab.communication.CommunicationHandler.open(CommunicationHandler.java:81)
                                                                      at org.fossasia.pslab.communication.ScienceLab.<init>(ScienceLab.java:81)
                                                                      at org.fossasia.pslab.others.ScienceLabCommon.openDevice(ScienceLabCommon.java:22)
                                                                      at org.fossasia.pslab.activity.MainActivity.onCreate(MainActivity.java:80)
                                                                      at android.app.Activity.performCreate(Activity.java:4492)
                                                                      at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1049)
                                                                      at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:1920)
                                                                      at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:1981) 
                                                                      at android.app.ActivityThread.access$600(ActivityThread.java:123) 
                                                                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1147) 
                                                                      at android.os.Handler.dispatchMessage(Handler.java:99) 
                                                                      at android.os.Looper.loop(Looper.java:137) 
                                                                      at android.app.ActivityThread.main(ActivityThread.java:4424) 
                                                                      at java.lang.reflect.Method.invokeNative(Native Method) 
                                                                      at java.lang.reflect.Method.invoke(Method.java:511) 
                                                                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:784) 
                                                                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:551) 
                                                                      at dalvik.system.NativeStart.main(Native Method) 
06-15 12:33:59.850 3203-3206/org.fossasia.pslab I/dalvikvm: threadid=3: reacting to signal 3
06-15 12:33:59.850 3203-3206/org.fossasia.pslab I/dalvikvm: Wrote stack traces to '/data/anr/traces.txt'
06-15 12:34:00.050 3203-3206/org.fossasia.pslab I/dalvikvm: threadid=3: reacting to signal 3
06-15 12:34:00.050 3203-3206/org.fossasia.pslab I/dalvikvm: Wrote stack traces to '/data/anr/traces.txt'
06-15 12:34:01.170 3203-3203/org.fossasia.pslab I/Process: Sending signal. PID: 3203 SIG: 9

```
</p></details>

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -5,7 +5,7 @@ android {
     buildToolsVersion "25.0.2"
     defaultConfig {
         applicationId "org.fossasia.pslab"
-        minSdkVersion 18
+        minSdkVersion 15
         targetSdkVersion 25
         versionCode 1
         versionName "1.0"

--- a/app/src/main/java/org/fossasia/pslab/communication/CommunicationHandler.java
+++ b/app/src/main/java/org/fossasia/pslab/communication/CommunicationHandler.java
@@ -6,6 +6,7 @@ import android.hardware.usb.UsbDeviceConnection;
 import android.hardware.usb.UsbEndpoint;
 import android.hardware.usb.UsbInterface;
 import android.hardware.usb.UsbManager;
+import android.os.Build;
 import android.os.SystemClock;
 import android.util.Log;
 
@@ -140,6 +141,9 @@ public class CommunicationHandler {
     }
 
     public int write(byte[] src, int timeoutMillis) throws IOException {
+        if (Build.VERSION.SDK_INT < 18) {
+            return writeSupportAPI(src, timeoutMillis);
+        }
         int written = 0;
         while (written < src.length) {
             int writeLength, amtWritten;
@@ -156,6 +160,35 @@ public class CommunicationHandler {
         }
         return written;
     }
+
+
+    // For supporting devices with API version < 18
+    public int writeSupportAPI(byte[] src, int timeoutMillis) throws IOException {
+        int written = 0;
+        while (written < src.length) {
+            final int writeLength;
+            final int amtWritten;
+            synchronized (mWriteBufferLock) {
+                final byte[] writeBuffer;
+                writeLength = Math.min(src.length - written, mWriteBuffer.length);
+                if (written == 0) {
+                    writeBuffer = src;
+                } else {
+                    // bulkTransfer does not support offsets for API level < 18, so make a copy.
+                    System.arraycopy(src, written, mWriteBuffer, 0, writeLength);
+                    writeBuffer = mWriteBuffer;
+                }
+                amtWritten = mConnection.bulkTransfer(mWriteEndpoint, writeBuffer, writeLength, timeoutMillis);
+            }
+            if (amtWritten <= 0) {
+                throw new IOException("Error writing " + writeLength
+                        + " bytes at offset " + written + " length=" + src.length);
+            }
+            written += amtWritten;
+        }
+        return written;
+    }
+
 
     public void clear() {
         mConnection.bulkTransfer(mReadEndpoint, mReadBuffer, 100, 50);

